File: problems/944/experiments/sixreg/diag_unlock.cpp
// Diagonal-partner criticality-feasibility, C++ port (split-INDEPENDENT form).
// unlock(w): B-w has a proper 3-colouring with some colour absent from the six
// cut units (units on deleted w removed). [T_p u T_l = all 6 units, split-free.]
// B criticality-feasible <=> every vertex w unlocks. Count feasible always-bal B.
// usage: geng -c -D6 n e:e | diag_unlock n
#include <cstdio>
#include <cstdint>
#include <cstring>
#include <string>
#include <vector>
#include <algorithm>
#include <iostream>
using namespace std;
static int N;
static const int MAXN = 32;
struct G { uint32_t adj[MAXN]; int deg(int u) const { return __builtin_popcount(adj[u]); } };
static int Bdef[MAXN];

// enumerate proper 3-colourings of G minus 'skip' (skip<0 = none); callback cb(col[]).
// cb returns true to stop early.
template<class CB>
static bool eachCol(const G& g, int skip, CB cb) {
  int order[MAXN], m = 0;
  for (int v = 0; v < N; v++) if (v != skip) order[m++] = v;
  for (int i = 1; i < m; i++){int k=order[i],j=i-1;while(j>=0&&g.deg(order[j])<g.deg(k)){order[j+1]=order[j];j--;}order[j+1]=k;}
  int8_t col[MAXN]; memset(col,-1,sizeof(col));
  int8_t tryc[MAXN]; memset(tryc,0,sizeof(tryc));
  int pos=0;
  while(pos>=0){
    if(pos==m){ if(cb(col)) return true; pos--; if(pos>=0)col[order[pos]]=-1; continue; }
    int x=order[pos]; bool adv=false;
    for(int c=tryc[pos];c<3;c++){
      bool ok=true; uint32_t nb=g.adj[x];
      while(nb){int u=__builtin_ctz(nb);nb&=nb-1; if(u==skip)continue; if(col[u]==c){ok=false;break;}}
      if(ok){col[x]=c;tryc[pos]=c+1;pos++;if(pos<m)tryc[pos]=0;adv=true;break;}
    }
    if(!adv){col[x]=-1;tryc[pos]=0;pos--;if(pos>=0)col[order[pos]]=-1;}
  }
  return false;
}
static bool alwaysBalanced(const G& g){
  bool any=false;
  bool bad = eachCol(g,-1,[&](const int8_t*col){
    int w[3]={0,0,0}; for(int v=0;v<N;v++) w[col[v]]+=Bdef[v];
    any=true;
    return !(w[0]==2&&w[1]==2&&w[2]==2);   // stop (true) if a colouring is unbalanced
  });
  return any && !bad;
}
// w unlocks: some colouring of g-w has a colour absent from the 6 units (minus w's units)
static bool unlocks(const G& g, int w, const vector<int>& units){
  return eachCol(g,w,[&](const int8_t*col){
    int present[3]={0,0,0};
    for(int v:units){ if(v==w)continue; present[col[v]]=1; }
    return !(present[0]&&present[1]&&present[2]);  // stop(true) if some colour missing
  });
}
static bool g6decode(const string& line, G& g){
  if(line.empty())return false; int n=line[0]-63; if(n!=N)return false;
  memset(g.adj,0,sizeof(g.adj));
  int nbits=n*(n-1)/2,need=(nbits+5)/6; if((int)line.size()<1+need)return false;
  int bit=0;
  for(int j=1;j<n;j++)for(int i=0;i<j;i++){int byte=1+bit/6,off=5-bit%6;
    if((line[byte]-63)>>off&1){g.adj[i]|=1u<<j;g.adj[j]|=1u<<i;}bit++;}
  return true;
}
int main(int argc,char**argv){
  N=argc>1?atoi(argv[1]):12;
  string line; long long total=0,ab=0,feasible=0;
  while(getline(cin,line)){
    while(!line.empty()&&(line.back()=='\r'||line.back()=='\n'))line.pop_back();
    if(line.empty()||line[0]=='>')continue;
    G g; if(!g6decode(line,g))continue;
    total++;
    int sb=0; for(int v=0;v<N;v++){Bdef[v]=6-g.deg(v);sb+=Bdef[v];} if(sb!=6)continue;
    if(!alwaysBalanced(g))continue;
    ab++;
    vector<int> units; for(int v=0;v<N;v++) for(int k=0;k<Bdef[v];k++) units.push_back(v);
    bool allUnlock=true;
    for(int w=0;w<N;w++){ if(!unlocks(g,w,units)){allUnlock=false;break;} }
    if(allUnlock){ feasible++; printf("FEASIBLE g6=%s\n",line.c_str()); }
  }
  printf("n=%d total=%lld alwaysBalanced=%lld criticalityFeasible=%lld\n",N,total,ab,feasible);
  return 0;
}

File: problems/944/experiments/sixreg/diag_unlock.py
"""Diagonal-partner criticality-unlock test (S-B2 vs S-K tension).
For an always-balanced shore-shaped B (conn, Delta<=6, e=3n-3, sum b=6, 3-col,
every proper colouring has cut-weight vector (2,2,2)):
  pick a 3+3 split of the six cut UNITS (vertex v contributes b[v] units) into
  T_p, T_l. A diagonal partner A forces p=l=c; G-w is 3-colourable iff B-w has a
  proper colouring with a colour c absent from BOTH T_p and T_l restricted to
  units not on w.  B is criticality-FEASIBLE for this split iff EVERY w in B
  unlocks.  Question: does ANY (B, split) achieve all-vertex unlock?
Reads g6 on stdin (any graphs); filters to always-balanced; tests all splits.
usage: geng -c -D6 n e:e | python diag_unlock.py n
"""
import sys, itertools
n = int(sys.argv[1])
def g6decode(s):
    nn = ord(s[0]) - 63
    adj = [set() for _ in range(nn)]
    bit = 0
    for j in range(1, nn):
        for i in range(j):
            byte = 1 + bit // 6; off = 5 - bit % 6
            if (ord(s[byte]) - 63) >> off & 1:
                adj[i].add(j); adj[j].add(i)
            bit += 1
    return nn, adj
def all_colourings(nn, adj, skip=-1):
    verts = [v for v in range(nn) if v != skip]
    verts.sort(key=lambda v: -len(adj[v]))
    col = {}
    def bt(i):
        if i == len(verts):
            yield dict(col); return
        v = verts[i]
        used = {col[u] for u in adj[v] if u in col}
        for c in range(3):
            if c in used: continue
            col[v] = c
            yield from bt(i+1)
            del col[v]
    yield from bt(0)
def always_balanced(nn, adj, b):
    any_col = False
    for col in all_colourings(nn, adj):
        w = [0,0,0]
        for v in range(nn): w[col[v]] += b[v]
        if tuple(sorted(w)) != (2,2,2): return False
        any_col = True
    return any_col
checked = 0; ab = 0; feasible = 0
feas_examples = []
for line in sys.stdin:
    line = line.strip()
    if not line or line[0] == '>': continue
    nn, adj = g6decode(line)
    b = [6 - len(adj[v]) for v in range(nn)]
    if sum(b) != 6: continue
    # connected
    seen={0}; st=[0]
    while st:
        x=st.pop()
        for y in adj[x]:
            if y not in seen: seen.add(y); st.append(y)
    if len(seen)!=nn: continue
    checked += 1
    if not always_balanced(nn, adj, b): continue
    ab += 1
    # units: list of vertices, with multiplicity b[v]
    units = []
    for v in range(nn):
        units += [v]*b[v]
    # precompute, for each w, all colourings of B-w (store per-vertex colour maps)
    cols_minus = {}
    for w in range(nn):
        cols_minus[w] = list(all_colourings(nn, adj, skip=w))
    # try every 3+3 split of the 6 unit-slots (indices into units)
    slot_idx = list(range(6))
    good_split = None
    seen_splits = set()
    for Tp_slots in itertools.combinations(slot_idx, 3):
        Tp = tuple(sorted(units[i] for i in Tp_slots))
        Tl = tuple(sorted(units[i] for i in slot_idx if i not in Tp_slots))
        key = (Tp, Tl) if Tp <= Tl else (Tl, Tp)
        if key in seen_splits: continue
        seen_splits.add(key)
        # check all w unlock
        all_unlock = True
        for w in range(nn):
            unlocked = False
            for col in cols_minus[w]:
                # union colours of Tp and Tl, units on w removed
                present = [False,False,False]
                for v in list(Tp)+list(Tl):
                    if v == w: continue   # w deleted -> its units gone
                    present[col[v]] = True
                if not all(present):   # some colour c missing from union
                    unlocked = True; break
            if not unlocked:
                all_unlock = False; break
        if all_unlock:
            good_split = (Tp, Tl); break
    if good_split:
        feasible += 1
        if len(feas_examples) < 5:
            feas_examples.append((line, good_split))
print(f"n={n} checked={checked} alwaysBalanced={ab} criticalityFeasible={feasible}")
for g, sp in feas_examples:
    print(f"  FEASIBLE g6={g} split={sp}")
